### PR TITLE
Guest OS hypervisor mapping improvements

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR includes the following Guest OS hypervisor mapping improvements.
- Support for the Guest OS / hypervisor mapping addition from upgrade path (using GuestOsMapper)
- Support to copy hypervisor mappings from older version from upgrade path (using GuestOsMapper)
- Provision to validate the OS mapping name in hypervisor when enabled using the new parameter 'osmappingcheckenabled' in addGuestOsMapping API
- Provision to get/check the hypervisor guest OS names (that can be used to add proper mapping) for VMware and XenServer using new API: getHypervisorGuestOsNames.
- Display the Guest OS hypervisor mappings in UI.

Fixes: #3609 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Manually tested add/update guest os mapping with the os name mapping check enabled, and tested the new getHypervisorGuestOsNames API with VMware and XenServer hypervisor.

```
(testenv-vmware) 🐱 > add guestosmapping hypervisor='VMware' hypervisorversion='5.5' osnameforhypervisor='centos' ostypeid=d6ea0862-0654-11ed-9169-1e00c10002c7 forced=true osmappingcheckenabled=true
🙈 Error: (HTTP 530, error code 4250) No host exists with hypervisor: VMware and version: 5.5, please specify available hypervisor and version
(testenv-vmware) 🐱 > add guestosmapping hypervisor='VMware' hypervisorversion='6.7.3' osnameforhypervisor='centos' ostypeid=d6ea0862-0654-11ed-9169-1e00c10002c7 forced=true osmappingcheckenabled=true
🙈 Error: (HTTP 530, error code 4250) Invalid hypervisor os mapping: centos for guest os: CentOS 7, hypervisor: VMware and version: 6.7.3
(testenv-vmware) 🐱 > get hypervisorguestosnames hypervisor='VMware' hypervisorversion='6.7.3' keyword='cent'
{
  "hypervisorguestosnames": {
    "guestoses": [
      {
        "osdisplayname": "Microsoft Windows Server 2003 Datacenter (64-bit)",
        "osnameforhypervisor": "winNetDatacenter64Guest"
      },
      {
        "osdisplayname": "Microsoft Windows Server 2003 Datacenter (32-bit)",
        "osnameforhypervisor": "winNetDatacenterGuest"
      },
      {
        "osdisplayname": "CentOS 8 (64-bit)",
        "osnameforhypervisor": "centos8_64Guest"
      },
      {
        "osdisplayname": "CentOS 7 (64-bit)",
        "osnameforhypervisor": "centos7_64Guest"
      },
      {
        "osdisplayname": "CentOS 6 (64-bit)",
        "osnameforhypervisor": "centos6_64Guest"
      },
      {
        "osdisplayname": "CentOS 6 (32-bit)",
        "osnameforhypervisor": "centos6Guest"
      },
      {
        "osdisplayname": "CentOS 4/5 or later (64-bit)",
        "osnameforhypervisor": "centos64Guest"
      },
      {
        "osdisplayname": "CentOS 4/5 or later (32-bit)",
        "osnameforhypervisor": "centosGuest"
      }
    ],
    "guestosescount": 8,
    "hypervisor": "VMware",
    "hypervisorversion": "6.7.3"
  }
}
(testenv-vmware) 🐱 > get hypervisorguestosnames hypervisor='VMware' hypervisorversion='6.7.3' keyword='core'
{
  "hypervisorguestosnames": {
    "guestoses": [
      {
        "osdisplayname": "CoreOS Linux (64-bit)",
        "osnameforhypervisor": "coreos64Guest"
      }
    ],
    "guestosescount": 1,
    "hypervisor": "VMware",
    "hypervisorversion": "6.7.3"
  }
}

(testenv-xen) 🐱 >get hypervisorguestosnames hypervisor='XenServer' hypervisorversion='7.1.0' keyword='core'
{
  "hypervisorguestosnames": {
    "guestoses": [
      {
        "osdisplayname": "CoreOS",
        "osnameforhypervisor": "CoreOS"
      }
    ],
    "guestosescount": 1,
    "hypervisor": "XenServer",
    "hypervisorversion": "7.1.0"
  }
}
(testenv-xen) 🐱 > get hypervisorguestosnames hypervisor='XenServer' hypervisorversion='7.1.0' keyword='cent'
{
  "hypervisorguestosnames": {
    "guestoses": [
      {
        "osdisplayname": "CentOS 4.8 (32-bit)",
        "osnameforhypervisor": "CentOS 4.8 (32-bit)"
      },
      {
        "osdisplayname": "CentOS 5 (32-bit)",
        "osnameforhypervisor": "CentOS 5 (32-bit)"
      },
      {
        "osdisplayname": "CentOS 7",
        "osnameforhypervisor": "CentOS 7"
      },
      {
        "osdisplayname": "CentOS 6 (64-bit)",
        "osnameforhypervisor": "CentOS 6 (64-bit)"
      },
      {
        "osdisplayname": "CentOS 4.6 (32-bit)",
        "osnameforhypervisor": "CentOS 4.6 (32-bit)"
      },
      {
        "osdisplayname": "CentOS 5 (64-bit)",
        "osnameforhypervisor": "CentOS 5 (64-bit)"
      },
      {
        "osdisplayname": "CentOS 4.7 (32-bit)",
        "osnameforhypervisor": "CentOS 4.7 (32-bit)"
      },
      {
        "osdisplayname": "CentOS 6 (32-bit)",
        "osnameforhypervisor": "CentOS 6 (32-bit)"
      },
      {
        "osdisplayname": "CentOS 4.5 (32-bit)",
        "osnameforhypervisor": "CentOS 4.5 (32-bit)"
      }
    ],
    "guestosescount": 9,
    "hypervisor": "XenServer",
    "hypervisorversion": "7.1.0"
  }
}
(testenv-xen) 🐱 > get hypervisorguestosnames hypervisor='XenServer' hypervisorversion='6.1.0'
{
  "accountid": "aa2fc411-0654-11ed-8c69-1e00b50002e7",
  "cmd": "org.apache.cloudstack.api.command.admin.guest.GetHypervisorGuestOsNamesCmd",
  "completed": "2022-07-28T09:28:12+0000",
  "created": "2022-07-28T09:28:12+0000",
  "jobid": "e5189994-d441-42d3-bb75-26647210a7c1",
  "jobprocstatus": 0,
  "jobresult": {
    "errorcode": 530,
    "errortext": "No host exists with hypervisor: XenServer and version: 6.1.0, please specify available hypervisor and version"
  },
  "jobresultcode": 530,
  "jobresulttype": "object",
  "jobstatus": 2,
  "userid": "aa310e7a-0654-11ed-8c69-1e00b50002e7"
}
```

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test matrix

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan help

**Comment by DaanHoogland** [comment]
I see nothing wrong with the code (even though sonar cloud does) . I have a question; how does this work with upgrade updates?

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java [hunk 1]
```diff
@@ -2831,6 +2894,14 @@ public GuestOSHypervisor updateGuestOsMapping(final UpdateGuestOsMappingCmd cmd)
             throw new InvalidParameterValueException("Unable to modify system defined Guest OS mapping");
         }
 
+        if (cmd.getOsMappingCheckEnabled()) {
+            GuestOS guestOs = ApiDBUtils.findGuestOSById(guestOsHypervisorHandle.getGuestOsId());
+            if (guestOs == null) {
+                throw new InvalidParameterValueException("Unable to find the guest OS for the mapping");
+            }
+            checkGuestOSHypervisorMapping(HypervisorType.getType(guestOsHypervisorHandle.getHypervisorType()), guestOsHypervisorHandle.getHypervisorVersion(), guestOs.getDisplayName(), osNameForHypervisor);
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/ManagementServerImpl.java:2903]: extra method

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
closing in favour of #6979

**Pull request closed without merging**